packages/Meshfree/src/DTK_MovingLeastSquaresOperator_def.hpp

```diff
@@ -0,0 +1,133 @@
+        DeviceType>::transformSourceCoordinates( source_points, _offset,
+                                                 target_points );
+    target_points = Kokkos::View<double **, DeviceType>( "empty" );
+
+    // Build P (vandermonde matrix)
+    // P is a single 1D storage for multiple P_i matrices. Each matrix is of
+    // size (#source_points_for_specific_target_point, basis_size)
+    auto p =
+        Details::MovingLeastSquaresOperatorImpl<DeviceType>::computeVandermonde(
+            source_points, PolynomialBasis() );
+
+    // Build phi (weight matrix)
+    auto phi =
+        Details::MovingLeastSquaresOperatorImpl<DeviceType>::computeWeights(
+            source_points, CompactlySupportedRadialBasisFunction( radius ) );
+
+    // Build A (moment matrix)
+    auto a =
+        Details::MovingLeastSquaresOperatorImpl<DeviceType>::computeMoments(
+            _offset, p, phi );
+
+    // TODO: it is computationally unnecessary to compute the pseudo-inverse as
+    // MxM (U*E^+*V) as it will later be just used to do MxV. We could instead
+    // return the (U,E^+,V) and do the MxV multiplication. But for now, it's OK.
+    auto t = Details::MovingLeastSquaresOperatorImpl<DeviceType>::invertMoments(
+        a, PolynomialBasis::size() );
+    auto inv_a = std::get<0>( t );
+    auto num_underdetermined = std::get<1>( t );
+
+    if ( num_underdetermined )
+        throw DataTransferKitException(
+            "Encountered the situation where "
+            "some of the systems are underdetermined. This could be a result "
+            "of the either very few source points found in the radius of a "
+            "target points (this can probably be fixed by increasing the "
+            "radius), or b) points lie on a single plane (this can be fixed "
+            "by reducing the basis order)." );
+
+    // NOTE: This assumes that the polynomial basis evaluated at {0,0,0} is
+    // going to be [1, 0, 0, ..., 0]^T.
+    _coeffs = Details::MovingLeastSquaresOperatorImpl<
+        DeviceType>::computePolynomialCoefficients( _offset, inv_a, p, phi,
+                                                    PolynomialBasis::size() );
+}
+
+template <typename DeviceType, typename CompactlySupportedRadialBasisFunction,
+          typename PolynomialBasis>
+void MovingLeastSquaresOperator<
+    DeviceType, CompactlySupportedRadialBasisFunction, PolynomialBasis>::
+    apply( Kokkos::View<double const *, DeviceType> source_values,
+           Kokkos::View<double *, DeviceType> target_values ) const
+{
+    // Retrieve values for all source points
+    source_values = Details::NearestNeighborOperatorImpl<DeviceType>::fetch(
+        _comm, _ranks, _indices, source_values );
+
+    // Apply A-1 (P^T phi)
+    auto new_target_values = Details::MovingLeastSquaresOperatorImpl<
+        DeviceType>::computeTargetValues( _offset, _coeffs, source_values );
+
+    Kokkos::deep_copy( target_values, new_target_values );
```

Comment:
Because `computeTargetValues()` returns the target values for clarity on what are in/out parameters.  But you are correct, it could in principle be passed as an argument.